src/core/algorithms/od/fastod/fastod.cpp

```diff
@@ -0,0 +1,301 @@
+#include "fastod.h"
+
+#include <memory>
+
+#include <boost/unordered/unordered_map.hpp>
+#include <easylogging++.h>
+
+#include "config/tabular_data/input_table/option.h"
+#include "config/time_limit/option.h"
+
+namespace algos {
+
+Fastod::Fastod() : Algorithm({}) {
+    PrepareOptions();
+}
+
+bool Fastod::IsTimeUp() const {
+    return time_limit_seconds_ > 0 && timer_.GetElapsedSeconds() >= time_limit_seconds_;
+}
+
+void Fastod::CCPut(AttributeSet const& key, AttributeSet attribute_set) {
+    cc_[key] = std::move(attribute_set);
+}
+
+fastod::AttributeSet const& Fastod::CCGet(AttributeSet const& key) {
+    return cc_[key];
+}
+
+void Fastod::PrepareOptions() {
+    RegisterOptions();
+    MakeLoadOptionsAvailable();
+}
+
+void Fastod::RegisterOptions() {
+    RegisterOption(config::kTableOpt(&input_table_));
+    RegisterOption(config::kTimeLimitSecondsOpt(&time_limit_seconds_));
+}
+
+void Fastod::MakeLoadOptionsAvailable() {
+    MakeOptionsAvailable({config::kTableOpt.GetName()});
+}
+
+void Fastod::MakeExecuteOptsAvailable() {
+    MakeOptionsAvailable({config::kTimeLimitSecondsOpt.GetName()});
+}
+
+void Fastod::LoadDataInternal() {
+    data_ = std::make_shared<DataFrame>(DataFrame::FromInputTable(input_table_));
+}
+
+void Fastod::ResetState() {
+    is_complete_ = false;
+
+    level_ = 1;
+    od_count_ = 0;
+    fd_count_ = 0;
+    ocd_count_ = 0;
+
+    result_asc_.clear();
+    result_desc_.clear();
+    result_simple_.clear();
+    context_in_current_level_.clear();
+    cc_.clear();
+    cs_asc_.clear();
+    cs_desc_.clear();
+
+    timer_ = Timer();
+    partition_cache_.Clear();
+}
+
+unsigned long long Fastod::ExecuteInternal() {
+    auto const start_time = std::chrono::system_clock::now();
+    auto const [ods_asc, ods_desc, ods_simple] = Discover();
+
+    auto const elapsed_milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::system_clock::now() - start_time);
```

Comment:
After changing `Discover()` signature can use `util::TimedInvoke`:
```suggestion
auto const elapsed_milliseconds = util::TimedInvoke(&Fastod::Discover, this);
```